tests/cxx/expl_inst_select.cc

```diff
@@ -0,0 +1,62 @@
+//===--- expl_inst_select.cc - test input file for iwyu -------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// This checks deduplication/selection of explicit template instantiations.
+
+// IWYU_ARGS: -I .
+
+#include "tests/cxx/expl_inst_select-d1.h"
+
+// An explicit instantiation definition anchors a prior declaration.
+// IWYU: Template is...*expl_inst_select-i1.h
+// IWYU: Template is...*expl_inst_select-i2.h.*for explicit instantiation
```

Comment:
@kimgr, do you know any reason to include an explicit instantiation declaration here?

Reply:
It seems to be covering the reporting driven from here:
https://github.com/include-what-you-use/include-what-you-use/blob/53ce70ecc5ee84899ca067f26ddce5cd77dcb064/iwyu.cc#L3263

... which as far as I can tell was added to solve:
https://github.com/include-what-you-use/include-what-you-use/issues/558

via https://github.com/include-what-you-use/include-what-you-use/pull/578.

The bug report in 558 stated that the explicit instantiation declaration is required for the definition, so presumably that's why? There may be some misunderstanding along the way there, of course.

Reply:
> The bug report in 558 stated that the explicit instantiation declaration is required for the definition

I don't see that. The original bug as-reported is just that IWYU produced some garbage on explicit instantiations.

> 3. The tool should not be suggesting to remove explicit template instantiations in the first place.

Agreed, but it should not do it regardless of the presence of the instantiation definition somewhere in the translation unit.

From [your comment](https://github.com/include-what-you-use/include-what-you-use/pull/578#issuecomment-421732597):
> Explicit template instantiation definition of `Template<short>`, right? That should require `template-decl-short.h`, otherwise it will create a reusable instantiation of the wrong template.

I've failed to understand, could you explain please? As far as I know, instantiation declarations are to suppress instantiations in a translation unit, and instantiation definitions are to the opposite purpose: to instantiate regardless of being used or not. Thus, they are mutually exclusive in essence, and definitions "beat" declarations, in fact. Consider:
```cpp
// tpl.h

template <typename T>
struct Tpl {
    static int i;
    int local_i;
    static int getStatic() { return i; }
    int getLocal() const { return local_i; }
};

// tpl.cpp

#include "tpl.h"

template <typename T>
int Tpl<T>::i = 2;

template class Tpl<int>;

// main.cpp

#include "tpl.h"

extern template class Tpl<int>;

int main() {
    Tpl<int> t{5};
    return t.getLocal() + t.getStatic();
}

```
This compiles, links and runs fine. Moreover, [clang LLVM IR output](https://godbolt.org/z/ensnfb744) and [GCC assembly output](https://godbolt.org/z/YjbzEPnK3) are the same with and without the instantiation declaration in the `tpl.cpp` translation unit.

An explicit instantiation definition requires the class template definition to be available (likewise an explicit instantiation declaration) but not the declaration, as seemed to me.

Reply:
> > Explicit template instantiation definition of Template<short>, right? That should require template-decl-short.h,
> > otherwise it will create a reusable instantiation of the wrong template.
>
> I've failed to understand, could you explain please?

I think I confused myself as mentioned in the next comment: https://github.com/include-what-you-use/include-what-you-use/pull/578#issuecomment-421734997.

If `Template<short>` was a _specialization_, I think that sentence would make a bit more sense. Let's say:

* There is an explicit specialization for `Template<short>` in `template-decl-short.h`
* Explicit instantiation definition `Template<short>` would instantiate the base template or the specialization, depending on what was visible at the time of definition

But that does seem like a separate constraint altogether.